samples/Microsoft.Bot.Samples.EchoBot-ASPNetCore-DI/Startup.cs

```diff
@@ -0,0 +1,77 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Bot.Builder;
+using Microsoft.Bot.Builder.Adapters;
+using Microsoft.Bot.Builder.BotFramework;
+using Microsoft.Bot.Builder.Middleware;
+using Microsoft.Bot.Builder.Storage;
+using Microsoft.Bot.Connector.Authentication;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.Bot.Samples.EchoBot_ASPNetCore_DI
+{
+    public class Startup
+    {
+        public Startup(IConfiguration configuration)
+        {
+            Configuration = configuration;
+        }
+
+        public IConfiguration Configuration { get; }
+
+        // This method gets called by the runtime. Use this method to add services to the container.
+        public void ConfigureServices(IServiceCollection services)
+        {
+            // Register required services.
+            services.AddSingleton<IConfiguration>(this.Configuration);
```

Comment:
This is... unexpected, in the way it works. 

The previous DI example had a cleaner approach for this.

Reply:
This is the way DI is done in the newer ASPNetCore (you can use Unity or AutoFac also but this is inbuilt into ASPNetCore). Hence I added an example showcasing. More documentation can be found at https://docs.microsoft.com/en-us/aspnet/core/fundamentals/dependency-injection
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/middleware/?tabs=aspnetcore2x

Reply:
I'm familiar with the ASP.NET DI and Middleware model. You can see what Tom has done here:
https://github.com/Microsoft/botbuilder-dotnet/blob/master/samples/AlarmBot/Startup.cs

Reply:
Hmm, this model will make it hard to reuse instances. For example if both UserStateManager and ConversationStateManager were to use same instance of IStorage it will require either instance of IStorage to be directly instantiated once and then put into container or declare it as singleton into the container and then use serviceProvider to get it's instance. This also makes writing end to end tests a bit hard as you have to replace the full adapter instance instead of being able to add or remove one single middleware (or httpclient for that matter) since they are not coming directly from the container.

Another issue with having single instance in the container is that it takes away the capability to have different scopes on different objects. For example out of 5 middlewares registered in a middlewareset one is scoped and rest 4 are singleton. This ideally should mean that the specific middleware, middlewareset and botframeworkadapter should be transient while everything else should be singleton. But having single instance in container means that every single instance will need to be transient which means unnecessary init of things like storage which can be pretty costly.

These are the reasons (if I invest more time I can probably come up with more) I went with registering every usable instance in the container and then let ASPNetCore's DI engine take care of the rest. Still if you feel this is too weird let me know and I will remove the sample code.

Reply:
I'm fine with that being how people build Bots, but we're trying really hard to keep the samples as simple as possible. Now, if something we're doing is PREVENTING this from working, then we should totally fix it. Short of that though, I would err on the side of "Whatever is the most simple", at least for now.

Reply:
I completely agree with your point of keeping the samples simple so that people can understand and grasp the concepts properly. At the same time I wanted to write a complex bot so that people have a guidance on how to utilize the power of these relatively simple concepts in complex situations (the reason I think this is useful is because as Tom mentioned that people tend to follow the path that is laid in examples, hence examples should show a better way) . The first thing that came obviously lacking from the sample was that everything is singleton, so it portrays a picture that middlewares need to be singleton otherwise the new SDK won't work. This is the gap I was trying to fill. Sorry if it made the whole thing look way too complex. 

Here's what I can think of, I will remove this sample if it seems that it is trying to solve a relatively simple mechanic using a relatively complex approach. Let me know if you want me to add a complex sample to showcase complex DI usage so as to provide a guidance for complex scenarios.